pkg/kf/internal/genericcli/xargs.go

```diff
@@ -0,0 +1,263 @@
+			ctx := cmd.Context()
+
+			if err := p.ValidateSpaceTargeted(); err != nil {
+				return err
+			}
+
+			logger := logging.FromContext(ctx)
+
+			if flags.allSpaces {
+				logger.Infof("Xargs %s in all spaces", options.PluralFriendlyName())
+			} else if t.Namespaced() {
+				logger.Infof("Xargs %s in space: %s", options.PluralFriendlyName(), p.Space)
+			}
+
+			commandTmpl := strings.Join(args, " ")
+			tmpl, err := template.New("command").Parse(commandTmpl)
+			if err != nil {
+				return fmt.Errorf("couldn't parse command template %s: %v", commandTmpl, err)
+			}
+
+			labelSelector := labels.Set{}
+			for k, v := range labelSelectorFlags {
+				if v != nil && *v != "" {
+					labelSelector[k] = *v
+				}
+			}
+
+			listOptions := metav1.ListOptions{
+				LabelSelector: labels.
+					SelectorFromSet(labelSelector).
+					Add(options.LabelRequirements()...).
+					String(),
+			}
+
+			// Collect spaces to target
+			var spaces []string
+			if flags.allSpaces {
+				spacesList, err := spacesClient.List(ctx)
+				if err != nil {
+					return err
+				}
+				for _, space := range spacesList {
+					spaces = append(spaces, space.Name)
+				}
+			} else {
+				spaces = strings.Split(p.Space, ",")
+			}
+
+			group, groupCtx := errgroup.WithContext(ctx)
+			limit := semaphore.NewWeighted(int64(flags.spaceConcurrency))
+
+			for _, space := range spaces {
+				limit.Acquire(groupCtx, 1)
+				if groupCtx.Err() != nil {
+					break
+				}
+
+				// Shadow space as a local to be captured by the closure
+				space := space
+				group.Go(func() error {
+					defer limit.Release(1)
+					return doSpace(groupCtx, cmd, t, listOptions, flags, space, tmpl)
+				})
+			}
+
+			if err := group.Wait(); err != nil {
+				return err
+			}
+
+			if flags.dryRun {
+				fmt.Fprintf(
+					cmd.OutOrStdout(), color.New(color.FgHiYellow).Sprint("Run with --dry-run=false to apply."))
+				fmt.Fprintln(cmd.OutOrStdout())
+			}
+
+			return nil
+		},
+	}
+
+	for flag, label := range options.LabelFilters() {
+		usage := fmt.Sprintf("Sets a filter for the %q label.", label)
+		labelSelectorFlags[label] = cmd.Flags().String(flag, "", usage)
+	}
+
+	flags.Add(cmd)
+
+	return cmd
+}
+
+type XargFlags struct {
+	dryRun           bool
+	allSpaces        bool
+	spaceConcurrency int
+	appConcurrency   int
+}
+
+func (f *XargFlags) Add(cmd *cobra.Command) {
+	cmd.Flags().BoolVar(&f.dryRun, "dry-run", true, "Enables dry-run mode, commands are printed but will not be executed.")
+	cmd.Flags().BoolVar(&f.allSpaces, "all-spaces", false, "Enables targeting all spaces in the cluster.")
+	cmd.Flags().IntVar(&f.spaceConcurrency, "space-concurrency", 1, "Number of spaces that may be operated on in parallel. Only meaningful when multiple spaces are targeted.")
+	cmd.Flags().IntVar(&f.appConcurrency, "app-concurrency", 1, "Number of apps within a space that may be operated on in parallel. Total concurrency will be upto space-concurrency * app-concurrency.")
```

Comment:
I think this should be "resource-concurrency"

I'm not sure how we'd generalize this command to also work on non-namespace scoped items, probably worth thinking about whether the flags are the same or different there or the right way to name them to indicate what's going on?

Reply:
Renamed to `resource-concurrency`, for the sake of simplicity and consistency I think using the same flag name irrespective of the actual resource type may be fine.